File: csharp/glp_link/primitives/LinkRuntime.cs
using GlpRuntime.Link.Reliability;
using GlpRuntime.Link.Seam;
using GlpRuntime.Runtime;

namespace GlpRuntime.Link.Primitives;

/// <summary>
/// The per-engine link-layer state the <c>'_link_setup'</c> kernel operates on
/// (feature 025, T030): the scheme→leaf <see cref="TransportRegistry"/>, the
/// idempotent LinkId→<see cref="LinkHandle"/> <see cref="LinkRegistry"/>, and the
/// single <see cref="LinkPump"/> that bridges asynchronous inbound frames onto the
/// runner thread (Option B).
/// </summary>
/// <remarks>
/// One <see cref="LinkRuntime"/> per <see cref="GlpRuntimeEngine"/>. It lives in the
/// hand-authored <c>glp_link</c> package, so the dependency flows
/// <c>glp_link → out/csharp</c> only (FR-057): the engine never references the link
/// layer; the link layer is wired in at the composition root (a test, or the REPL's
/// boot path) by <see cref="LinkKernels.Install"/>, which registers the kernels and
/// hands the engine a pump to drain.
/// </remarks>
public sealed class LinkRuntime
{
    /// <summary>Scheme→transport-leaf selection (FR-013); register leaves before setup.</summary>
    public TransportRegistry Transports { get; } = new();

    /// <summary>
    /// Scheme→payload-codec selection (feature 050, FR-005). Register a codec for a scheme at the
    /// composition root (the <c>"quic"</c> crdtmsg codec); any unregistered scheme resolves to the
    /// default ground-relay blob, so loopback/tcp are unchanged. The selected codec is stamped onto
    /// each <see cref="LinkHandle"/> at establishment.
    /// </summary>
    public PayloadCodecRegistry PayloadCodecs { get; } = new();

    /// <summary>Idempotent-at-identity LinkId→handle registry (FR-007).</summary>
    public LinkRegistry Links { get; } = new();

    /// <summary>
    /// Distributed-GC coordinator (T024, FR-024). Each established link registers its
    /// reclamation hooks here (the registry entry today; send-registry goals + reply-table
    /// entries as those subsystems gain live per-link state); <c>'_link_close'</c> and the
    /// graceful stream-end <c>[]</c> close run them via <c>LinkClose.Teardown</c> so the
    /// runtime returns to its pre-link baseline (T035). One per instance; not thread-safe.
    /// </summary>
    public LinkReclaimer Reclaimer { get; } = new();

    /// <summary>The one inbound pump this engine's run-to-quiescence driver services (Option B).</summary>
    public LinkPump Pump { get; }

    /// <summary>
    /// Accepted-but-not-yet-adopted inbound connections, keyed by the ground
    /// <see cref="LinkId"/> carried in the request token (T033 Option A). Populated by
    /// <c>'_link_listen'</c> when it reads an in-band request frame; drained by
    /// <c>'_link_accept'</c>, which adopts the endpoint into <see cref="Links"/> via the
    /// shared establish-core. The two-step split (listen produces requests; accept
    /// establishes one) is the ratified separation of concerns.
    /// </summary>
    public Dictionary<LinkId, ILinkEndpoint> Pending { get; } = new();

    /// <summary>The engine whose heap the kernel binds and whose goal queue the pump feeds.</summary>
    public GlpRuntimeEngine Engine { get; }

    public LinkRuntime(GlpRuntimeEngine engine)
    {
        Engine = engine ?? throw new ArgumentNullException(nameof(engine));
        Pump = new LinkPump(engine);
    }
}

File: glp_runtime/lib/link/primitives/link_runtime.dart
import '../../runtime/runtime.dart';
import '../reliability/link_reclaimer.dart';
import '../seam/i_link_endpoint.dart';
import '../seam/link_id.dart';
import 'link_pump.dart';
import 'link_registry.dart';
import 'transport_registry.dart';

/// The per-engine link-layer state the `'_link_setup'` kernel operates on
/// (feature 025, T030): the scheme→leaf [TransportRegistry], the
/// idempotent LinkId→[LinkHandle] [LinkRegistry], and the
/// single [LinkPump] that bridges asynchronous inbound frames onto the
/// runner thread (Option B).
///
/// One [LinkRuntime] per [GlpRuntime]. It lives in the
/// hand-authored `glp_link` package, so the dependency flows
/// `glp_link → out/csharp` only (FR-057): the engine never references the link
/// layer; the link layer is wired in at the composition root (a test, or the REPL's
/// boot path) by `LinkKernels.install`, which registers the kernels and
/// hands the engine a pump to drain.
class LinkRuntime {
  /// Scheme→transport-leaf selection (FR-013); register leaves before setup.
  final TransportRegistry transports = TransportRegistry();

  /// Idempotent-at-identity LinkId→handle registry (FR-007).
  final LinkRegistry links = LinkRegistry();

  /// Distributed-GC coordinator (T024, FR-024). Each established link registers its
  /// reclamation hooks here (the registry entry today; send-registry goals + reply-table
  /// entries as those subsystems gain live per-link state); `'_link_close'` and the
  /// graceful stream-end `[]` close run them via `LinkClose.Teardown` so the
  /// runtime returns to its pre-link baseline (T035). One per instance; not thread-safe.
  final LinkReclaimer reclaimer = LinkReclaimer();

  /// The one inbound pump this engine's run-to-quiescence driver services (Option B).
  late final LinkPump pump;

  /// Accepted-but-not-yet-adopted inbound connections, keyed by the ground
  /// [LinkId] carried in the request token (T033 Option A). Populated by
  /// `'_link_listen'` when it reads an in-band request frame; drained by
  /// `'_link_accept'`, which adopts the endpoint into [links] via the
  /// shared establish-core. The two-step split (listen produces requests; accept
  /// establishes one) is the ratified separation of concerns.
  final Map<LinkId, ILinkEndpoint> pending = <LinkId, ILinkEndpoint>{};

  /// The engine whose heap the kernel binds and whose goal queue the pump feeds.
  final GlpRuntime engine;

  LinkRuntime(GlpRuntime engine)
      // ignore: unnecessary_null_comparison, prefer_initializing_formals
      : engine = engine {
    // ignore: unnecessary_null_comparison, dead_code
    if (engine == null) throw ArgumentError.notNull('engine');
    pump = LinkPump(engine);
  }

  /// Release every link-layer OS resource (feature 025 clean shutdown): cancel the
  /// pump's background recv loops (completing its cancel signal unparks each
  /// `await recvBytes`), then dispose every established and every parked-but-unadopted
  /// transport endpoint — close + destroy the socket and cancel its reader
  /// subscription. After this the isolate holds no link sockets through the registry,
  /// so a one-shot REPL / embedded host can terminate promptly instead of being held
  /// alive by parked recv Futures (which is what hung the back-to-back link suite's
  /// `wait`). Best-effort and idempotent; a still-pending transport `listen`/`connect`
  /// future (its socket not yet in the registry) is the REPL's `exit(0)` backstop.
  Future<void> dispose() async {
    pump.dispose();
    for (final handle in links.handles) {
      final ep = handle.endpoint;
      if (ep != null) {
        try {
          await ep.dispose();
        } on Object {
          /* best-effort: process is shutting down */
        }
      }
    }
    for (final ep in pending.values) {
      try {
        await ep.dispose();
      } on Object {
        /* best-effort */
      }
    }
    pending.clear();
  }
}
